### ISSUE TITLE
Program is rejected as unsafe depending on if an imported module is checked beforehand

### ISSUE THREAD

**Issue by RyanGlScott** [ucsd-progsys/liquidhaskell#1112]
Consider these two modules, the latter of which imports the former:

```haskell
{-@ LIQUID "--higherorder"        @-}
{-@ LIQUID "--exactdc"            @-}
module Prod where

{-@ data Product f g p = Product { prd1 :: (f p), prd2 :: (g p) } @-}
data Product f g p = Product (f p) (g p)

{-@ axiomatize leqProd @-}
leqProd :: Eq (f p)
        => (f p -> f p -> Bool) -> (g p -> g p -> Bool)
        -> Product f g p -> Product f g p -> Bool
leqProd leqFP leqGP (Product x1 y1) (Product x2 y2) =
  if x1 == x2
    then leqGP y1 y2
    else leqFP x1 x2
{-# INLINE leqProd #-}
```
```haskell
{-@ LIQUID "--higherorder"        @-}
{-@ LIQUID "--exactdc"            @-}
module Foo where

import Prod
```

To ensure you can reproduce these results, delete `.liquid` if it exists in the current directory. Now try running `liquid` on `Foo`:

```
$ /opt/cabal/head/bin/cabal new-run liquid -- Foo.hs 
Up to date
LiquidHaskell Copyright 2013-17 Regents of the University of California. All Rights Reserved.


**** DONE:  A-Normalization ****************************************************
 

**** DONE:  Extracted Core using GHC *******************************************
 

**** DONE:  Transformed Core ***************************************************
 

**** DONE:  Uniqify & Rename ***************************************************
 
Working -9223372036854775808% []

**** DONE:  annotate ***********************************************************
 

**** RESULT: SAFE **************************************************************
```

All is well. Now delete `.liquid` again. This time, check `Prod` before checking `Foo` (without deleting `.liquid` in between):

```
$ /opt/cabal/head/bin/cabal new-run liquid -- Prod.hs 
Up to date
LiquidHaskell Copyright 2013-17 Regents of the University of California. All Rights Reserved.


**** DONE:  A-Normalization ****************************************************
 

**** DONE:  Extracted Core using GHC *******************************************
 

**** DONE:  Transformed Core ***************************************************
 

**** DONE:  Uniqify & Rename ***************************************************
 
Working -9223372036854775808% []

**** DONE:  annotate ***********************************************************
 

**** RESULT: SAFE **************************************************************
$ /opt/cabal/head/bin/cabal new-run liquid -- Foo.hs 
Up to date
LiquidHaskell Copyright 2013-17 Regents of the University of California. All Rights Reserved.


**** DONE:  A-Normalization ****************************************************
 

**** DONE:  annotate ***********************************************************
 

**** RESULT: ERROR *************************************************************


 /home/rgscott/Documents/Hacking/Haskell/verified-instances/generic-proofs/Prod.hs:8:16: Error: GHC Error
  
 8 | {-@ axiomatize leqProd @-}
                    ^
  
     Not in scope: variable or data constructor ` Prod.prd1 '
```

This time, `Foo` is rejected!

**Comment by ranjitjhala** [comment]
Ok keep this open because I really ought to add an error message.

1. Running `foo` does not actually trigger any checks on `Prod` which is why its safe. When there are imports and such, you need to run liquid on the dependencies first. (Its not like `ghc --make`; running the dependencies is a "separate concern". 

2. When run with `--exact-dc` we  now require EITHER that you just write the haskell definition (thats it!)

```
data Product f g p = Product { prd1 :: (f p), prd2 :: (g p) } 
```

or if you must write a refined data-type, that fields be *defined* in the Haskell definition:

```
{-@ data Product f g p = Product { prd1 :: (f p), prd2 :: (g p) } @-}
data Product f g p = Product { prd1 :: (f p), prd2 :: (g p) } 
```

The latter is only needed e.g. if there are some dependencies you want to capture 
on/between fields.

In short, just *delete* the line `{-@ data ... @-}` and this should work.

**Comment by RyanGlScott** [comment]
Keep in mind that I don't _want_ to use the records `pr1` and `pr2` in the definition of `Product`—I only put them there originally because LH demanded it. Will LH accept proofs over a Haskell data definition of `data Product f g p = Product (f p) (g p)` (with no refinement)?

Also, it's quite annoying that `liquid` _doesn't_ act like `ghc --make`. Currently, I test files by running a script which invokes `liquid` one-by-one on each of the Haskell files in a directory, which obviously ignores import ordering. This means that in order to accommodate the way `liquid` works, I have to hardcode in the script the exact order in which to run `liquid` on each file so as to avoid the issue described above. It would thus be extremely convenient if `liquid` could figure this out for me.

**Comment by ranjitjhala** [comment]
> Will LH accept proofs over a Haskell data 
> definition of `data Product f g p = Product (f p) (g p)` (with no refinement)?

Yes. If for some reason you _want_ the field-accessor 
names `pr1` and `pr2` e.g. to refer to them in 
refinements or whatever, then you need to explicitly 
declare them **both** in the GHC definition and the 
Liquid definition. Otherwise, you can just omit the 
Liquid definition altogether. So the following works:

```
{-@ LIQUID "--higherorder"  @-}
{-@ LIQUID "--exactdc" @-}
module Prod where

data Product f g p = Product (f p) (g p)

{-@ axiomatize leqProd @-}
leqProd :: Eq (f p)
        => (f p -> f p -> Bool) -> (g p -> g p -> Bool)
        -> Product f g p -> Product f g p -> Bool
leqProd leqFP leqGP (Product x1 y1) (Product x2 y2) =
  if x1 == x2
    then leqGP y1 y2
    else leqFP x1 x2
{-# INLINE leqProd #-}
```

**Comment by RyanGlScott** [comment]
Let me phrase my question slightly differently then. Let's suppose I have a module `ProdDef`:

```haskell
module ProdDef where

data Product f g p = Product (f p) (g p) deriving Eq
```

This is in a package I have no control over, so I can't run `liquid` on this file. From another module, I import `ProdDef`:

```haskell
{-@ LIQUID "--higherorder"        @-}
{-@ LIQUID "--exactdc"            @-}
module Prod where

import ProdDef

{-@ axiomatize leqProd @-}
leqProd :: Eq (f p)
        => (f p -> f p -> Bool) -> (g p -> g p -> Bool)
        -> Product f g p -> Product f g p -> Bool
leqProd leqFP leqGP (Product x1 y1) (Product x2 y2) =
  if x1 == x2
    then leqGP y1 y2
    else leqFP x1 x2
{-# INLINE leqProd #-}
```

If I just run `liquid` on this, LH is unhappy:

```
$ /opt/cabal/head/bin/cabal new-run liquid -- Prod.hs 
Up to date
LiquidHaskell Copyright 2013-17 Regents of the University of California. All Rights Reserved.


**** DONE:  A-Normalization ****************************************************
 

**** DONE:  annotate ***********************************************************
 

**** RESULT: ERROR *************************************************************


 /home/rgscott/Documents/Hacking/Haskell/verified-instances/generic-proofs/Prod.hs:7:16: Error: Illegal type specification for `Prod.leqProd`
  
 7 | {-@ axiomatize leqProd @-}
                    ^
  
     Prod.leqProd :: (Eq a b) =>
                     lq2:(a b -> a b -> Bool) -> lq3:(c b -> c b -> Bool) -> lq4:(Product a c b) -> lq5:(Product a c b) -> {VV : Bool | VV == Prod.leqProd lq2 lq3 lq4 lq5
                                                                                                                                        && VV == (if lqdc##$select lq4 == lqdc##$select lq5 then lq3 (lqdc##$select lq4) (lqdc##$select lq5) else lq2 (lqdc##$select lq4) (lqdc##$select lq5))}
     Sort Error in Refinement: {VV : bool | (VV == Prod.leqProd lq2 lq3 lq4 lq5
                                             && VV == (if lqdc##$select##ProdDef.Product##1 lq4 == lqdc##$select##ProdDef.Product##1 lq5 then lq3 (lqdc##$select##ProdDef.Product##2 lq4) (lqdc##$select##ProdDef.Product##2 lq5) else lq2 (lqdc##$select##ProdDef.Product##1 lq4) (lqdc##$select##ProdDef.Product##1 lq5)))}
     Unbound Symbol lqdc##$select##ProdDef.Product##1
 Perhaps you meant: lqdc##$select##GHC.Tuple.(,,)##1, lqdc##$select##GHC.Tuple.(,)##1, lqdc##$select##GHC.Tuple.(,,,)##1
```

So clearly the answer here isn't "just delete the line `{-@ data ... @-}`". But what?

**Comment by ranjitjhala** [comment]
No you cannot do this, and I can't think of any way we could _possibly_ do this, because `liquid` needs access to the source to "lift" the data definition in the "exact" style.

Can you add a line:

```haskell
{-@ data Product @-} 
``` 

to the `Prod` module? I wonder if that will work ...

**Comment by RyanGlScott** [comment]
I've updated `Prod` to have `{-@ data Product @-}` as you suggested:

```haskell
module ProdDef where

data Product f g p = Product (f p) (g p) deriving Eq
```
```haskell
{-@ LIQUID "--higherorder"        @-}
{-@ LIQUID "--exactdc"            @-}
module Prod where

import ProdDef

{-@ data Product @-}

{-@ axiomatize leqProd @-}
leqProd :: Eq (f p)
        => (f p -> f p -> Bool) -> (g p -> g p -> Bool)
        -> Product f g p -> Product f g p -> Bool
leqProd leqFP leqGP (Product x1 y1) (Product x2 y2) =
  if x1 == x2
    then leqGP y1 y2
    else leqFP x1 x2
{-# INLINE leqProd #-}
```

LH still screams when checking `Prod`:

```
$ /opt/cabal/head/bin/cabal new-run liquid -- Prod.hs 
Up to date
LiquidHaskell Copyright 2013-17 Regents of the University of California. All Rights Reserved.


**** DONE:  A-Normalization ****************************************************
 

**** DONE:  annotate ***********************************************************
 

**** RESULT: ERROR *************************************************************


 /home/rgscott/Documents/Hacking/Haskell/verified-instances/generic-proofs/Prod.hs:9:16: Error: Illegal type specification for `Prod.leqProd`
  
 9 | {-@ axiomatize leqProd @-}
                    ^
  
     Prod.leqProd :: (Eq a b) =>
                     lq2:(a b -> a b -> Bool) -> lq3:(c b -> c b -> Bool) -> lq4:(Product a c b) -> lq5:(Product a c b) -> {VV : Bool | VV == Prod.leqProd lq2 lq3 lq4 lq5
                                                                                                                                        && VV == (if lqdc##$select lq4 == lqdc##$select lq5 then lq3 (lqdc##$select lq4) (lqdc##$select lq5) else lq2 (lqdc##$select lq4) (lqdc##$select lq5))}
     Sort Error in Refinement: {VV : bool | (VV == Prod.leqProd lq2 lq3 lq4 lq5
                                             && VV == (if lqdc##$select##ProdDef.Product##1 lq4 == lqdc##$select##ProdDef.Product##1 lq5 then lq3 (lqdc##$select##ProdDef.Product##2 lq4) (lqdc##$select##ProdDef.Product##2 lq5) else lq2 (lqdc##$select##ProdDef.Product##1 lq4) (lqdc##$select##ProdDef.Product##1 lq5)))}
     Unbound Symbol lqdc##$select##ProdDef.Product##1
 Perhaps you meant: lqdc##$select##GHC.Tuple.(,,)##1, lqdc##$select##GHC.Tuple.(,)##1, lqdc##$select##GHC.Tuple.(,,,)##1
```

**Comment by ranjitjhala** [comment]
Got it, will fix this ASAP!

**Comment by ranjitjhala** [comment]
Am working on this, but a simple way to get the above working, as is, 
is to pass the `--exact` flag at the command line. So:

```
$ stack exec -- liquid --exact ProdDef.hs 
$ stack exec -- liquid --exact Prod.hs
```

**Comment by RyanGlScott** [comment]
OK, but this isn't really a viable workaround since, again, I don't have the ability to run `liquid` on the `ProdDef.hs` analog in my actual use case. (Moreover, I tried running `liquid --exact` on `Prod.hs` alone, but experienced the same error.)

**Comment by ranjitjhala** [comment]
Actually @RyanGlScott can you let me know if why 
(other than the make/dependency hassle) the above 
i.e. (passing in --exact as a command line param) 
will not work for you?

I fear that allowing "reflecting" to happen when processing 
some _other_ module (e.g. here, reflecting the definition of 
`Product` -- defined in ProdDef while processing a different 
module `Prod`) is likely to open weird cans of worms, as there 
may not be a unique place that a datatype got reflected from...

**Comment by ranjitjhala** [comment]
Why can't you run `liquid` on `ProdDef` in your real use case again?

**Comment by RyanGlScott** [comment]
To be clear: the issue isn't `--exact`. It's the fact that you have to run `liquid` on `ProdDef.hs` _at all_. I only included here for the purposes of having an easily digestible example, but in actuality, `Product` is located in a module of some other package (`base`) that I don't have the ability to reflect in its definition site. Therefore, the ability to reflect `Product` in another location is crucial.

> I fear that allowing "reflecting" to happen when processing
> some other module (e.g. here, reflecting the definition of
> `Product` -- defined in ProdDef while processing a different
> module `Prod`) is likely to open weird cans of worms, as there
> may not be a unique place that a datatype got reflected from...

Sure. These are definitely _orphan_ refinements, so I imagine LH would need some kind of orphan refinement consistency check (much like GHC has for orphan instances) to make this work. But IIRC, LH already errors if it sees duplicate refinements for a datatype, so perhaps this already happens?

**Comment by ranjitjhala** [comment]
Hmm. Would you be ok with, say, having to add explicit annotations in `Prod.hs` that force lifting, e.g.

```
{-@ external data Product @-}
```

?

**Comment by RyanGlScott** [comment]
Absolutely! I'm willing to do some extra legwork to account for the fact that I'm doing reflection outside of a data type's definition site. I want to reflect this thing by hook or by crook.

**Comment by ranjitjhala** [comment]
@RyanGlScott this should be fixed in `develop` now, see 

tests/import/lib/T1112.hs

and 

tests/import/lib/T1112Lib.hs

**Comment by ranjitjhala** [comment]
Also no legwork needed: `--exact-data-cons` will just automatically reflect the relevant `TyCon`.
